### DISCUSSION TITLE
Common format for image operations with Libvips

### DISCUSSION THREAD

**Discussion by frederico-apolonia** [libvips/libvips#3789] [Q&A]
Hello,

We have designed an image selector component which functions as follows:
1. Grabs the correct image from a file storage
2. Sends it to Libvips to apply image operations (crop, resize, image compression, images composites)
3. Returns image to client
![imageselector](https://github.com/libvips/libvips/assets/8880155/8410c2ed-0a5a-4a12-b8ce-c1411c7c53c0)

As we process a very large number of images on a daily basis, we're looking to optimise the image processing component to return smaller images whilst processing them quicker. The stored images don't follow a pattern format, they could be PNG, JPEG, GIF and we're testing if a common format between all images would speed up Libvips. So far we've found out that using the `.vips` and `.tiff` formats helps with Libvips processing time, but it comes with a high penalty of increasing the file storage around 15x.

We've also tried using `.avif`, `.jxl`, `.webp` and `.heic`. The first three reduced the footprint on the file storage and created smaller final images, but processing times increased when compared with our baseline (image service without a common format).

This is leaving us with the following questions:
1. Is there any way to optimise `.vips` images? From my understanding it is saving the decompressed dataclass image, are there parameters to create a smaller file?
2. Besides using `.vips` or `.tiff` image formats, would Libvips be quicker with a common format? If so, is there any format that you'd recommend us trying?

Thank you

**Comment by jcupitt** [comment]
Hello @frederico-apolonia,

It's very hard to optimise a system without a clear benchmark.  I would take a three step approach:

1. Develop a small, complete, standalone benchmark you can share which accurately represents your use case ... docker containers are a convenient way to build these things in a mostly reproducible way
2. Set a goal for the optimisation (eg. "reduce memory use by 25%", or "get end-to-end latency under 20ms")
3. Profile, investigate, and explore the solution space!

For example, as you say, processing time depends on the file format, but also very strongly on the exact libvips pipeline you need to execute, the network architecture, the available compute and IO resources, the system config, and so on. Unless you have a concrete thing to test, you won't know which factors are important and which can be ignored.

**Reply by frederico-apolonia** [reply]
Hi @jcupitt , thank you for your quick answer and sorry for the delay!

I have developed a small Python script to benchmark the idea and try to get some answers, [base_format_test.py.zip](https://github.com/libvips/libvips/files/13739741/base_format_test.py.zip).

The results that we obtained were clear, we get better results with `.vips` and `.tiff` but the file sizes are very large. Is there any optimisation that you're aware of to make `.vips` files smaller? Are there parts of the file that can be removed to reduce it's file size?

Thank you

**Reply by jcupitt** [reply]
vips format is just a huge array -- it's quick because there's no compression. There's not much you can do there.

The fastest general format is tiled jpeg tiff.  It combines small size (the jpeg compression) with random access (tiled). If you use the pyramidal variant you can do very fast shrinking as well. It can't do transparency though, and it's lossy.

I'll have a look at your script and see if there's anything that can be improved there.

**Reply by frederico-apolonia** [reply]
Thank you about the explanation about the vips and tiff formats! Seems like they could be out of the equation for our use-case then.

**Comment by jcupitt** [comment]
I looked at your test script quickly and thought of a few things:

- You are just timing individual operations. This is a very slow way to use libvips -- instead of exposing single operations, I would code up common tasks (eg. downsize + colour correct + sharpen + write as mime jpeg could be one task) as complete pipelines and time them. You'll see a *huge* speedup.

- You are not using sequential mode, so you're not getting any image streaming. You can put `access="sequential"` on most of your `new_from_file` for a large speedup and drop in memory use.

- You are not using `new_from_source`. libvips can read and write directly to and from S3 buckets, which would give another large drop in latency.

- You are using `resize` rather than `thumbnail`. There's another factor of up to maybe x4 speedup here.

**Reply by frederico-apolonia** [reply]
Thanks for taking a look into the code. 

I'll investigate how I could pipeline the Libvips tasks, it could also be very useful for our prod deployment.
I'm not understanding where I'm not pipelining the tasks tho. Given for example the `apply_resize_op`:
```python
def apply_resize_op(x, elem, img_name):
    base_vips_image = pyvips.Image.new_from_file(f"{IMAGE_FORMAT_NAME}/{img_name}{get_image_format(img_name)}")
    base_vips_image = base_vips_image.resize(0.75)
    out_file = os.path.join(out_dir, f'{x}-{elem}{original_formats[img_name]}')
    base_vips_image.write_to_file(f'{out_file}[{qualities[original_formats[img_name]]["default"]}keep=none]')
    return out_file
```
Apart from not using `access="sequential"` and using `resize` I don't see where this isn't a pipeline. From what I'm understanding after reading `pyvips` documentation, this would be a pipeline (unless image access mode is what defines if it is or isn't a pipeline).

I'll also checkout the `new_from_file` and `resize` optimisations, thank you for the guidance.

As for the `new_from_source`, as this is a POC, I'm only reading from the disk - I might have not been very clear here - and not from a specific S3. Thanks for the heads-up.

**Reply by jcupitt** [reply]
Oh, no problem.

I just meant you could do more in one task. If you run:

```
disk -> resize -> disk -> crop -> disk -> compose -> disk
```

That'll be many, many times slower than:

```
disk -> resize -> crop -> compose -> disk
```

So it's best to make an API that implements complete tasks rather than just exposing a set of separate operations.

**Reply by frederico-apolonia** [reply]
Ah I see. That's what we're doing. Each iteration will only run 1 operation and save to disk. We do orchestrate many operations in prod, but we only read once and write once.

I have another question about how Libvips handles memory on C lib. We're opening the files via memory buffer after downloading them from S3 and only then opening them with `vips_image_new_from_buffer`. As you spoke about directly spoke about opening directly from S3, are we slowing down our system by handling image creation this way?

**Reply by jcupitt** [reply]
Yes, you can process directly from one S3 bucket to another, with no buffering in memory and no separate download/upload step. There were some introductory notes here:

https://www.libvips.org/2019/12/11/What's-new-in-8.9.html

**Reply by jcupitt** [reply]
> Ah I see. That's what we're doing. Each iteration will only run 1 operation and save to disk. 

That'll be very slow. I'll make you a demo prog, hang on.

**Reply by frederico-apolonia** [reply]
> Yes, you can process directly from one S3 bucket to another, with no buffering in memory and no separate download/upload step. There were some introductory notes here:
> 
> https://www.libvips.org/2019/12/11/What's-new-in-8.9.html

Ah that's promising. We're not uploading the image again to S3 so if I understood it correctly we'd see memory gains by letting vips manage the image download from S3.

> That'll be very slow. I'll make you a demo prog, hang on.

I might have not been clear (again ahah, sorry), on this POC each iteration will open an image from the disk, do the operation and write back to disk. It's not certain it'll open two times the same image. There could be an optimisation that I might be missing, but here's a snippet of the operations list:
```
resize-485
resize-403
quality-29
resize-196
resize-283
quality-92
resize-109
quality-35
compose-38;logo_5
resize-217
```
It's `operation`-`image` 

Thank you again for taking your time!

**Reply by jcupitt** [reply]
Heh now I'm completely confused. Why are you running a series of single operations? Can't you batch them up and do several operations in one go?

**Reply by frederico-apolonia** [reply]
I'm not doing several operations in one go because it wouldn't be a POC of our deployed system, where we receive multiple requests to process different images and then return them to our clients.

**Comment by jcupitt** [comment]
You can test thumbnail like this:

```
john@orange ~/pics $ vipsheader nina.jpg
nina.jpg: 6048x4032 uchar, 3 bands, srgb, jpegload
john@orange ~/pics $ time vips thumbnail nina.jpg x.jpg 605

real	0m0.085s
user	0m0.049s
sys	0m0.042s
john@orange ~/pics $ time vips resize nina.jpg x.jpg 0.1

real	0m0.144s
user	0m0.087s
sys	0m0.073s
```

So about twice as quick, for this image.

**Comment by jcupitt** [comment]
Perhaps:

```python
#!/usr/bin/env python3

import sys
import pyvips
import time

print("processing an operation at a time ...")
start = time.time()

image = pyvips.Image.thumbnail(sys.argv[1], 5000)
image.write_to_file("t1.jpg", Q=90)

small = pyvips.Image.new_from_file("t1.jpg")
image = pyvips.Image.new_from_file(sys.argv[1])
mage = image.composite2(small, "over")
image.write_to_file("t2.jpg", Q=90)

image = pyvips.Image.new_from_file("t2.jpg")
image.write_to_file("t3.jpg", Q=75)

end = time.time()
print(f"took {end - start:.2f}s")

print("processing as a single pieline ...")
start = time.time()

small = pyvips.Image.thumbnail(sys.argv[1], 5000)
image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")
image = image.composite2(small, "over")
image.write_to_file("t4.jpg", Q=75)

end = time.time()
print(f"took {end - start:.2f}s")
```

I see:

```
$ ./pipeline.py ~/pics/nina.jpg
processing an operation at a time ...
took 0.38s
processing as a single pieline ...
took 0.23s
```

So about 2x faster on this laptop.

If I split them out into separate programs you can see a big difference in memory use too:

```
$ /usr/bin/time -f %M:%e ./pipeline.py ~/pics/nina.jpg
processing an operation at a time ...
took 0.43s
224064:0.57
$ /usr/bin/time -f %M:%e ./pipeline.py ~/pics/nina.jpg
processing as a single pieline ...
took 0.23s
102136:0.35
```

About half of the memory.

**Reply by frederico-apolonia** [reply]
Yep, it makes total sense to me. You're taking full advantage of how vips works on the second approach, hence why it's faster.
